internal/experiment/runner.go

```diff
@@ -0,0 +1,291 @@
+				r.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to assemble application", activity.URL, err), activity.URL)
+				continue
+			}
+
+			assembledBytes, err := r.generateApp(*assembledApp)
+			if err != nil {
+				r.handleErrors(ctx, fmt.Errorf("%s (%s): %w", "failed to generate application", activity.URL, err), activity.URL)
+				continue
+			}
+
+			exp := &redskyv1beta2.Experiment{}
+			if err := yaml.Unmarshal(assembledBytes, exp); err != nil {
+				r.handleErrors(ctx, fmt.Errorf("%s: %w", "invalid experiment generated", err), activity.URL)
+				continue
+			}
+
+			switch activity.Tags[0] {
+			case applications.TagScan:
+				template, err := server.ClusterExperimentToAPITemplate(exp)
+				if err != nil {
+					r.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to convert experiment template", err), activity.URL)
+					continue
+				}
+
+				if err := r.apiClient.UpdateTemplate(ctx, templateURL, *template); err != nil {
+					r.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to save experiment template in server", err), activity.URL)
+					continue
+				}
+			case applications.TagRun:
+				// We wont compare existing scan with current scan
+				// so we can preserve changes via UI
+
+				// Get previous template
+				previousTemplate, err := r.apiClient.GetTemplate(ctx, templateURL)
+				if err != nil {
+					r.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to get experiment template from server", err), activity.URL)
+					continue
+				}
+
+				// Overwrite current scan results with previous scan results
+				if err = server.APITemplateToClusterExperiment(exp, &previousTemplate); err != nil {
+					r.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to convert experiment template", err), activity.URL)
+					continue
+				}
+
+				// At this point the experiment should be good to create/deploy/run
+				// so let's create all the resources and #profit
+
+				// Create additional RBAC ( primarily for setup task )
+				if err = r.createServiceAccount(ctx, assembledBytes); err != nil {
+					r.handleErrors(ctx, err, activity.URL)
+				}
+
+				if err = r.createClusterRole(ctx, assembledBytes); err != nil {
+					r.handleErrors(ctx, err, activity.URL)
+				}
+
+				if err = r.createClusterRoleBinding(ctx, assembledBytes); err != nil {
+					r.handleErrors(ctx, err, activity.URL)
+				}
+
+				// Create configmap for load test
+				if err = r.createConfigMap(ctx, assembledBytes); err != nil {
+					r.handleErrors(ctx, err, activity.URL)
+				}
+
+				if err = r.createExperiment(ctx, exp); err != nil {
+					r.handleErrors(ctx, err, activity.URL)
+				}
+			}
+
+			if err := r.apiClient.DeleteActivity(ctx, activity.URL); err != nil {
+				r.handleErrors(ctx, fmt.Errorf("%s: %w", "failed to delete activity", err), activity.URL)
+			}
+		}
+	}
+}
+
+func (r *Runner) handleErrors(ctx context.Context, err error, activityURL string) {
+	failure := applications.ActivityFailure{FailureMessage: err.Error()}
```

Comment:
This should have a reason code: generally the error checking looks for a non-empty reason code but doesn't look at the message.